File: C.Hash/274.H-Index/274.H-Index_Hash.cpp
class Solution {
public:
    int hIndex(vector<int>& citations) 
    {
        if (citations.size()==0) return 0;
        
        map<int,int>Map;
        for (int i=0; i<citations.size(); i++)
            Map[citations[i]]++;
        auto q = vector<pair<int,int>>(Map.begin(),Map.end());
        
        int sum=0;
        
        for (int i=q.size()-1; i>=0; i--)
        {
            if (q[i].first>sum+q[i].second)
                sum+=q[i].second;
            else if (q[i].first==sum+q[i].second)
                return q[i].first;
            else
                return max(q[i].first,sum);
        }
        
        return sum;        
    }
};

File: Hash/274.H-Index/274.H-Index_Hash.py
class Solution:
    def hIndex(self, citations):
        """
        :type citations: List[int]
        :rtype: int
        """
        if len(citations)==0: return 0
        if len(citations)==1: return min(citations[0],1)
        counts = collections.Counter(citations)
        L = list(counts.items())
        L = sorted(L,key=lambda x:x[0],reverse=True)
        H_paper = 0
        print(L)
        for i in range(len(L)):                    
            if H_paper+L[i][1] == L[i][0]:
                return L[i][0]
            elif H_paper+L[i][1]>L[i][0]:
                return max(L[i][0],H_paper)
            else:
                H_paper+=L[i][1]
        return H_paper
                
                
        
        
